orttraining/orttraining/core/optimizer/sum_accumulate_transformer.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/graph/graph_utils.h"
+#include "orttraining/core/optimizer/sum_accumulate_transformer.h"
+
+using namespace ONNX_NAMESPACE;
+
+namespace onnxruntime {
+
+bool SumAccumulateTransformer::SatisfyCondition(const Node& node) const {
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Sum", {8, 13})) {
+    return false;
+  }
+
+  const auto next_node = node.OutputNodesBegin();
+
+  if (next_node != node.OutputNodesEnd() &&
+      node.GetOutputEdgesCount() == 1 &&
+      graph_utils::IsSupportedOptypeVersionAndDomain(*next_node, "InPlaceAccumulator", {1}, kMSDomain) &&
+      next_node->InputDefs().size() == 2 &&
+      graph_utils::GetNodeOutputName(node, 0) == graph_utils::GetNodeInputName(*next_node, 1)) {
+    return true;
+  }
+  return false;
+}
+
+Status SumAccumulateTransformer::ApplyImpl(Graph& graph, bool& modified, int /*graph_level*/, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_ids = graph_viewer.GetNodesInTopologicalOrder();
+
+  // Traverse backward from the bottom of the graph, so that the recompute nodes
+  // for lower layers are executed earlier
+  for (int i = static_cast<int>(node_ids.size() - 1); i >= 0; --i) {
```

Comment:
how about `for (auto node_id_it = node_ids.rbegin(); node_id_it != node_ids.rend(); ++node_id_it)` to avoid casting?